lxd/main_daemon.go

```diff
@@ -76,23 +76,42 @@ func (c *cmdDaemon) Run(cmd *cobra.Command, args []string) error {
 	chIgnore := make(chan os.Signal)
 	signal.Notify(chIgnore, unix.SIGHUP)
 
+	chForceShutdown := make(chan bool)
+	chDone := make(chan bool)
+
 	s := d.State()
 	select {
 	case sig := <-ch:
 		if sig == unix.SIGPWR {
-			logger.Infof("Received '%s signal', shutting down containers", sig)
+			logger.Infof("Received '%s signal', shutting down instances", sig)
 			d.Kill()
-			containersShutdown(s)
+			instancesShutdown(s)
 			networkShutdown(s)
 		} else {
 			logger.Infof("Received '%s signal', exiting", sig)
 			d.Kill()
 		}
 
 	case <-d.shutdownChan:
-		logger.Infof("Asked to shutdown by API, shutting down containers")
+		logger.Infof("Asked to shutdown by API, waiting for all operations to finish")
+		// Cancelling the context will make everyone aware that we're shutting down.
+		d.cancel()
+
+		go func() {
+			waitForOperations(s, chForceShutdown)
+			chDone <- true
+		}()
+
+		select {
+		case <-chDone:
+			// Wait for the running operations to stop
+		case <-d.shutdownChan:
+			// If we run `lxd shutdown` again, shut down forcefully.
+			chForceShutdown <- true
```

Comment:
@monstermunchkin can you use an empty struct channel here and just close it? that way it indicates you're not sending any data, just that the action of closing the channel means the force shut down event has occurred.